datalad/cmd.py

```diff
@@ -66,7 +66,49 @@ def __call__(self, cmd, *args, **kwargs):
         else:
             raise ValueError("Argument 'command' is neither a string nor a callable.")
 
-    def run(self, cmd, log_stdout=True, log_stderr=True):
+
+    def _get_output_online(self, proc, log_stderr, log_stdout, return_output=False):
+        stdout, stderr = [], []
+        while proc.poll() is None:
+            if log_stdout:
+                line = proc.stdout.readline()
+                if line != '':
+                    if return_output:
+                        stdout += line
+                    self.log("stdout| " + line.rstrip('\n'))
+                    # TODO: what level to log at? was: level=5
+                    # Changes on that should be properly adapted in
+                    # test.cmd.test_runner_log_stdout()
+            else:
+                pass
+
+            if log_stderr:
+                line = proc.stderr.readline()
+                if line != '':
+                    if return_output:
+                        stderr += line
+                    self.log("stderr| " + line.rstrip('\n'),
+                             level=logging.ERROR)
+                    # TODO: what's the proper log level here?
+                    # Changes on that should be properly adapted in
+                    # test.cmd.test_runner_log_stderr()
+            else:
+                pass
+
+        return stdout, stderr
+
+    def _get_output_communicate(self, proc, log_stderr, log_stdout, return_output=False):
+        """Delegate collection of output to proc.communicate.  It always collects
+        output, thus
+        """
+        out = proc.communicate()
+        for o, n in zip(out, ('stdout', 'stderr')):
+            if o:
+                self.log("%s| " % n + o.rstrip('\n'))
```

Comment:
Oh, right. Got it.

2015-03-12 14:00 GMT+01:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/cmd.py
> <https://github.com/bpoldrack/datalad/pull/1#discussion_r26298800>:
>
> > +                    # TODO: what's the proper log level here?
> > +                    # Changes on that should be properly adapted in
> > +                    # test.cmd.test_runner_log_stderr()
> > +            else:
> > +                pass
> > +
> > +        return stdout, stderr
> > +
> > +    def _get_output_communicate(self, proc, log_stderr, log_stdout, return_output=False):
> > +        """Delegate collection of output to proc.communicate.  It always collects
> > +        output, thus
> > +        """
> > +        out = proc.communicate()
> > +        for o, n in zip(out, ('stdout', 'stderr')):
> > +            if o:
> > +                self.log("%s| " % n + o.rstrip('\n'))
>
> whenever it is a multiline string i.e. with '\n' -- it would just dump it
> without prefixing with "stdout|" at each '\n'. Previuosly it would replace
> each '\n' with '|' prefix
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bpoldrack/datalad/pull/1/files#r26298800>.
>